### PR TITLE
fix(poller): fix remote server duplication (#11552)

### PR DESCRIPTION
**Pull request by kduret**
## Description

Cherry-pick of #11552 

**Fixes** MON-14579

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.04.x
- [x] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x (master)

### CODE REVIEW

File: src/Centreon/Application/ApiPlatform.php [hunk 1]
```diff
@@ -28,27 +28,27 @@
...
     private $version;
 
     /**
      * Get the API version
      *
-     * @return float
+     * @return string
      */
-    public function getVersion(): float
+    public function getVersion(): string
```
Comment (TamazC) [inline src/Centreon/Application/ApiPlatform.php:40]: The is no more need to keep type convertion in "PlatformTopologyRegisterRepositoryAPI" class private function getToken line 93

File: src/Centreon/Domain/PlatformInformation/Model/PlatformInformationFactory.php [hunk 2]
```diff
@@ -54,6 +54,9 @@ public function createRemoteInformation(array $information): PlatformInformation
         $platformInformation = new PlatformInformation($isRemote);
```
Comment (TamazC) [inline src/Centreon/Domain/PlatformInformation/Model/PlatformInformationFactory.php:54]:
```suggestion
        $platformInformation = new PlatformInformation(true);
```
And remove the line above `$isRemote = true;`

Comment (TamazC) [reply src/Centreon/Domain/PlatformInformation/Model/PlatformInformationFactory.php:54]: You could maybe refactor public function createCentralInformation in this way :
```
public function createCentralInformation(): PlatformInformation
    {
       return new PlatformInformation(false);
    }
```

File: src/Centreon/Domain/PlatformTopology/Model/PlatformRegistered.php [hunk 3]
```diff
@@ -194,14 +194,11 @@ public function setHostname(?string $hostname): PlatformInterface
      */
     private function checkIpAddress(?string $address): ?string
```
Comment (TamazC) [inline src/Centreon/Domain/PlatformTopology/Model/PlatformRegistered.php:195]: I think this method shouldn't return anything and it's a utility function

File: src/CentreonRemote/Application/Webservice/CentreonConfigurationRemote.php [hunk 4]
```diff
@@ -540,42 +542,54 @@ public function authorize($action, $user, $isInternal = false)
      * @param bool $noProxy to do not use configured proxy
      */
     private function addServerToListOfRemotes(
```
Comment (TamazC) [inline src/CentreonRemote/Application/Webservice/CentreonConfigurationRemote.php:544]: I think the name could be something like "addOrUpdateRemoteServer" and maybe you could split this method in 2 smaller methods : addRemoteServer and updateRemoteServer

File: src/CentreonRemote/Application/Webservice/CentreonRemoteServer.php [hunk 5]
```diff
@@ -146,21 +146,22 @@ public function postAddToWaitList(): string
...
-            )";
-
-        $insert = $this->pearDB->prepare($insertQuery);
-        $insert->bindValue(':ip', $ip, \PDO::PARAM_STR);
-        $insert->bindValue(':app_key', $appKey, \PDO::PARAM_STR);
-        $insert->bindValue(':version', $version, \PDO::PARAM_STR);
-        $insert->bindValue(':http_method', $httpScheme, \PDO::PARAM_STR);
-        $insert->bindValue(':http_port', $httpPort, \PDO::PARAM_INT);
-        $insert->bindValue(':no_check_certificate', $noCheckCertificate, \PDO::PARAM_STR);
         try {
```
Comment (TamazC) [inline src/CentreonRemote/Application/Webservice/CentreonRemoteServer.php:149]: It seems that this public method "postAddToWaitList" is never used.

File: www/include/configuration/configServers/DB-Func.php [hunk 6]
```diff
@@ -1494,3 +1496,73 @@ function ipCanBeUpdated(array $options): bool
...
+        );
+        $insertRemoteServerStatement->bindValue(
+            ":noCheckCertificate",
+            $remoteServerInformation["no_check_certificate"],
+            \PDO::PARAM_STR
+        );
+        $insertRemoteServerStatement->bindValue(":noProxy", $remoteServerInformation["no_proxy"], \PDO::PARAM_STR);
+        $insertRemoteServerStatement->bindValue(":serverId", $newId, \PDO::PARAM_INT);
+        $insertRemoteServerStatement->execute();
+    }
```
Comment (TamazC) [inline www/include/configuration/configServers/DB-Func.php:1567]:
```suggestion
function duplicateRemoteServerInformation(int $duplicatedId, int $newId): void
{
    global $pearDB;

    $serverInfo = getRemoteServerInformation($duplicatedId);

    if (empty($serverInfo)) {
        return;
    }

    $httpPort = $serverInfo['http_port'] !== null ? (int) $serverInfo['http_port'] : null;

    $statement = $pearDB->prepare(
        "INSERT INTO `remote_servers` (ip, `version`, is_connected,
        centreon_path, http_method, http_port, no_check_certificate, no_proxy, server_id) VALUES
        (:ip, :version, :isConnected, :centreonPath, :httpMethod, :httpPort,
        :noCheckCertificate, :noProxy, :serverId)"
    );
    $statement->bindValue(':ip', $serverInfo['ip'], \PDO::PARAM_STR);
    $statement->bindValue(':version', $serverInfo['version'], \PDO::PARAM_STR);
    $statement->bindValue(':isConnected', $serverInfo['is_connected'], \PDO::PARAM_INT);
    $statement->bindValue(':centreonPath', $serverInfo['centreon_path'], \PDO::PARAM_STR);
    $statement->bindValue(':httpMethod', $serverInfo['http_method'], \PDO::PARAM_STR);
    $statement->bindValue(':httpPort', $httpPort, \PDO::PARAM_INT);
    $statement->bindValue(':noCheckCertificate', $serverInfo['no_check_certificate'],\PDO::PARAM_STR);
    $statement->bindValue(':noProxy', $serverInfo['no_proxy'], \PDO::PARAM_STR);
    $statement->bindValue(':serverId', $newId, \PDO::PARAM_INT);
    $statement->execute();
}
```

File: www/install/php/Update-21.10.10.php [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        if ($server = $selectServerStatement->fetch()) {
+            $updateRemoteStatement->bindValue(':server_id', $server['id'], \PDO::PARAM_INT);
+            $updateRemoteStatement->bindValue(':id', $remoteId, \PDO::PARAM_INT);
+            $updateRemoteStatement->execute();
+        } else {
+            $deleteRemoteStatement->bindValue(':id', $remoteId, \PDO::PARAM_INT);
+            $deleteRemoteStatement->execute();
+        }
+    }
+}
```
Comment (TamazC) [inline www/install/php/Update-21.10.10.php:102]:
```suggestion
function migrateRemoteServerRelations(\CentreonDB $pearDB): void
{
    $processedIps = [];

    $selectStatement = $pearDB->prepare('SELECT id FROM nagios_server WHERE ns_ip_address = :ip_address');
    $deleteStatement = $pearDB->prepare('DELETE FROM remote_servers WHERE id = :id');
    $updateStatement = $pearDB->prepare('UPDATE remote_servers SET server_id = :server_id WHERE id = :id');

    $remoteServers = $pearDB->query('SELECT id, ip FROM remote_servers');

    foreach ($remoteServers->fetchAll(\PDO::FETCH_KEY_PAIR) as $id => $ip) {
        if (in_array($ip, $processedIps)) {
            $deleteStatement->bindValue(':id', $id, \PDO::PARAM_INT);
            $deleteStatement->execute();
        }

        $processedIps[] = $ip;

        $selectStatement->bindValue(':ip_address', $ip, \PDO::PARAM_STR);
        $selectStatement->execute();
        if ($server = $selectStatement->fetch()) {
            $updateStatement->bindValue(':server_id', $server['id'], \PDO::PARAM_INT);
            $updateStatement->bindValue(':id', $id, \PDO::PARAM_INT);
            $updateStatement->execute();
        } else {
            $deleteStatement->bindValue(':id', $id, \PDO::PARAM_INT);
            $deleteStatement->execute();
        }
    }
}
```

Comment (TamazC) [reply www/install/php/Update-21.10.10.php:102]: but you should test if usage of $remoteServers->fetchAll(\PDO::FETCH_KEY_PAIR) is correct

**Review by TamazC — APPROVED** [review]

**Review by jeremyjaouen — APPROVED** [review]

**Pull request merged**